// powerpala/Plugin.js
const { existsSync } = require('fs');
const { join } = require("path");
const { sleep, createElement } = require('powerpala/util');
const { resolveCompiler } = require('powerpala/compilers');

module.exports = class Plugin {
  constructor(){
    this.ready = false;
    this.styles = {};
  }

  get dependencies(){
    return this.manifest.dependencies;
  }

  get optionalDependencies(){
    return this.manifest.optionalDependencies;
  }

  get effectiveOptionalDependencies(){
    const deps = this.manifest.optionalDependencies;
    const disabled = powerpala.settings.get('disabledPlugins', []);
    return deps.filter(d => powerpala.pluginManager.get(d) !== void 0 && !disabled.includes(d));
  }

  get allDependencies(){
    return this.dependencies.concat(this.optionalDependencies);
  }

  get allEffectiveDependencies(){
    return this.dependencies.concat(this.effectiveOptionalDependencies);
  }

  get dependents(){
    const dependents = [ ...powerpala.pluginManager.plugins.values() ].filter(p => p.manifest.dependencies.includes(this.entityID));
    return [ ...new Set(dependents.map(d => d.entityID).concat(...dependents.map(d => d.dependents))) ];
  }

  get color(){
    return '#ff9600';
  }

  loadStylesheet(path){
    let resolvedPath = path;
    if(!existsSync(resolvedPath)){
      resolvedPath = join(powerpala.pluginManager.pluginDir, this.entityID, path);

      if (!existsSync(resolvedPath)) throw new Error(`Cannot find "${path}"! Make sure the file exists and try again.`);
    }

    const id = Math.random().toString(36).slice(2);
    const compiler = resolveCompiler(resolvedPath);
    const style = createElement('style', { id: `style-${this.entityID}-${id}`, 'data-powerpala': true, 'data-plugin': true });
    document.head.appendChild(style);

    this.styles[id] = compiler;
    const compile = async () => {
      style.innerHTML = await compiler.compile();
    };

    compiler.enableWatcher();
    compiler.on('src-update', compile);
    this[`__compileStylesheet_${id}`] = compile;
    this[`__compiler_${id}`] = compiler;
    return compile();
  }

  log (...data) {
    console.log(`%c[Powerpala:Plugin:${this.constructor.name}]`, `color: ${this.color}`, ...data);
  }

  debug (...data) {
    console.debug(`%c[Powerpala:Plugin:${this.constructor.name}]`, `color: ${this.color}`, ...data);
  }

  warn (...data) {
    console.warn(`%c[Powerpala:Plugin:${this.constructor.name}]`, `color: ${this.color}`, ...data);
  }

  error (...data) {
    console.error(`%c[Powerpala:Plugin:${this.constructor.name}]`, `color: ${this.color}`, ...data);
  }

  async _update(force=false){
    const success = await super._update(force);
    if (success && this.ready) await powerpala.pluginManager.remount(this.entityID);
    return success;
  }

  async _load(){
    try {
      while(!this.allEffectiveDependencies.every(pluginName => powerpala.pluginManager.get(pluginName).ready)) await sleep(1);
      if(this.startPlugin instanceof Function) await this.startPlugin();

      this.log('Plugin loaded');
      this.ready = true;
    } catch (e) {
      this.error('An error occurred during initialization!', e);
    }
  }

  async _unload(){
    try {
      for(const id in this.styles){
        this[`__compiler_${id}`].on('src-update', this[`__compileStylesheet_${id}`]);
        document.getElementById(`style-${this.entityID}-${id}`).remove();
        this.styles[id].disableWatcher();
      }
      if(this.stopPlugin instanceof Function) await this.stopPlugin();

      this.log('Plugin unloaded');
      this.ready = false;
    } catch (e) {
      this.error('An error occurred during shutting down! It\'s heavily recommended reloading Discord to ensure there are no conflicts.', e);
    }
  }
}
