// powerpala/Theme.js
const { resolveCompiler } = require('./compilers');
const { createElement } = require('./util');

module.exports = class Theme {
  constructor (themeID, manifest) {
    const styleManager = powerpala.styleManager;
    this.compiler = resolveCompiler(manifest.effectiveTheme);
    this.manifest = manifest;
    this.applied = false;
  }

  apply(){
    if(!this.applied){
      this.applied = true;
      const style = createElement('style', { id: `theme-${this.entityID}`, 'data-powerpala': true, 'data-theme': true });
      document.head.appendChild(style);
      this._doCompile = async () => {
        style.innerHTML = await this.compiler.compile();
      };

      this.compiler.enableWatcher();
      this.compiler.on('src-update', this._doCompile);
      return this._doCompile();
    }
  }

  remove () {
    if(this.applied){
      this.applied = false;
      this.compiler.off('src-update', this._doCompile);
      document.getElementById(`theme-${this.entityID}`).remove();
      this.compiler.disableWatcher();
    }
  }
}
